### DISCUSSION TITLE
Neural Amp Modeller with PiPedal

### DISCUSSION THREAD

**Discussion by rerdavies** [rerdavies/pipedal#96] [Tips and Tricks]
A set of changes that allow Neural Amp Modeler to be used with PiPedal have just been merged to https://github.com/mikeoliphant/neural-amp-modeler-lv2

This plugin allows models built by Neural Amp Modeler to be run in an LV2 PLugin. The results are impressive. And the ability to choose from thousands of downloadable  .NAM models on the internet makes this plugin a good replacement for the bundled TooB ML plugin.

A huge collection of Neural Amp models is available at [ToneHunt](https://tonehunt.org/).

CPU use is ferocious -- about 45%, but they do run stably without under-runs on PiPedal.  Neural Amp Modeler can load models with a variety of different architectures, some more costly, and some less costly. Reportedly, some Neural Amp models will have lower CPU use, but I'm not sure how to tell which are which. For the meantime, TooB ML provides better performance, even if it provides a much smaller set of models to choose from.

Mike and I might collaborate on building a Debian package in the near future. If anyone has experience building for multiple platforms and operating systems, I'd be grateful you could help me out. In the meantime, you'll have to build it yourself from source.

Because it's licensed as GPL, I can't bundle it with PiPedal's distribution. (I'm not actually sure I'm even allowed to load it under the terms of the GPL). In the meantime, you'll have to build and install it yourself. Building it is relatively straightforward if you follow the instructions on the GitHub page.  To install the plugin once you have built it, run:
```
    sudo cp  [your build directory here]/build/neural_amp_modeler.lv2/* /usr/local/lib/lv2/neural_amp_modeler.lv2/
```
In PiPedal, click on the _File_ control to upload .nam models using your web browser or the PiPedal Android client.

**Comment by 38github** [comment]
Thank you for your work on making it possible to use this LV2 plugin in PiPedal. I use it quite often and would very much like to make a little pedalboard with a Raspberry Pi or similar sized SBC and a small USB audio interface with acceptable Hi-Z input and outputs. Right now I use a Focusrite Scarlett 2i2 Gen2 and it works astonishly well using 64/3 latency. With Neural Amp Modeler I have to increase it to 128/3 though but that is probably because the CPU usage is 70-90%.

One thing though. Every time I go to the slot where Neural Amp Modeler is the web page says it says it either unable to load the model or something else. The audio is not interrupted or anything though. I also get a reconnect abitrarely when I change models but I think that is due to the low CPU power of the La Frite. I am considering buying a Orange Pi 5 but to me there is something about trying to make things work on "as little as needed".

**Reply by rerdavies** [reply]
A present for La Frite: I just uploaded a critical functional fix (v1.0.22) that... how to put it... makes the ToobML plugin sound as good as it should. tbh, I'm kind of surprised at how good it does sound. 

La Frite, on the other should find that it's considerably less demanding to run than Neural Amp Modeler. It's the same general principle, using much smaller models than the Neural Amp models usually are. The models may not be as good as the Neural Amp models, and you can't upload your own, but the models that are there are... really surprisingly pleasantly good.

**Comment by rerdavies** [comment]
You might want to make sure you have the latest version of neural amp
modeler from Mike Oliphant's page. There was a brief period where Neural
Amp modeller complained about not being able to find it's default impulse
file named "click.wav". And the latest and greatest PiPedal now displays
any error messages logged by plugins (it didn't before). I'm pretty sure
that's been fixed.

The current version, with pipedal fixes now merged:

     https://github.com/mikeoliphant/neural-amp-modeler-lv2


I'm not sure what the reconnect is about. It doesn't sound right to me.
Upgrade to the latest version, and see if that makes a difference. And if
you continue to see a disconnect, maybe send me log files, and let's check
that something more serious isn't going wrong.

To generate log files:

   journalctl -b 0 | grep pipedald > ~/logs.txt

and send me the result.

I'm please to see that the Orange Pi 5 is finally generally available! I've
been watching out for it since they first announced it last year. And the
price is good for what it is! Definitely serious competition for Raspberry
PI! Maybe twice the performance (50% for 2.4Ghz instead of 1.8ghz, and a
substantial additional improvement for an A-76 instead of an A-72), and two
or four times the RAM. The PiPedal build is starting to get a bit crowded
on an 8GB Pi.

One caveat: you should ask around to confirm that it will run an OS with a
kernel version of at least 5.15. There were substantial fixes to USB audio
in that release of the kernel, including major performance improvements..
Versions of Linux prior to that do not support USB audio well at all.  I'm
not certain, but I think your Focusrite Scarlet was one of the very few USB
audio cards that did work well on pre-5.15 Linux.

The relevant command to display the OS's kernel version is:

```
    > uname -a
    Linux machinename *5.15.0-71-lowlatency* #78-Ubuntu *SMP PREEMPT *Wed
Apr 19 12:17:25 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```
The SMB PREEMPT part matters too. It indicates a low-latency kernel.
Raspberry PI's come with an SMP PREEMPT kernel by default. Linux Mint
includes a low-latency kernel as part of the distribution, but doesn't
install it by default. It's not straightforward to find or build
low-latency kernels for other distros.

I matters because, traditionally, ARM Linux kernels were based on
corresponding Android kernel versions, which lagged far behind mainline
Linux kernels. Why? Because the chip venders don't really care about
writing drivers for Linuxen, but they care very much about making sure that
their chipsets will run Android. The real profits for ARM chipsets is on
phones. So they write drivers and firmware for Android, instead of current
mainstream Linux versions. I had my eye on a very promising Raspberry
Pi-alike for a while, only to find out that they had decided to ship with a
Linux 4.x kernel instead of a 5.15 kernel. Also a Rockchip-based product.
And the Armel-based multimedia boxes suffer particularly from this problem
as well.

This may have changed a bit now that ARM support is in upstream Linux. But
my understanding is that boot configuration has changed substantially
between Linux 4.x and Linux 5.x; and the boot changes make it difficult to
port Wifi and Bluetooth drivers, and (most difficult of all) video drivers
with copy-protection support for Netflix &c. These drivers are usually
provided by the chipset manufacturer. So manufacturers of micro-PCs get
stuck with whatever their chipset manufacturer decides to support.

**Reply by rerdavies** [reply]
Took a moment to look up benchmark results for Orange Pi 5. My guess: it should get ~3x the performance of a Raspberry Pi 4! Benchmarks show 3.25x median improvement, with notable failures to perform on benchmarks that I think are using low-power cores in addition to the high-performance cores to do multi-threaded Neural Net calculations (things will not go well if they do).

Unfortunately, they're using kernel version 5.10, I think.

**Reply by 38github** [reply]
Interesting. Would it be possible to change the PiPedal Systemd service to assign it to certain cores using cgroups?

**Reply by rerdavies** [reply]
It absolutely should be possible. Probably best to wait and see if it actually is a real problem that affects Pipedal out of the box. Very unfortunately, I think the 5.10 kernel is going to be a problem though. Older kernels don't properly support USB audio 1.1, except for a handful of ancient USB audio adapters that have been special-cased.  I'm not entirely sure when the major rewrite went in. Possibly in 5.10; but there are a bunch of bug fixes for USB audio in 5.13, and a major reduction in USB audio latency in 5.14.

**Comment by 38github** [comment]
I think I know, from looking at journalctl, what the reason might be... OOM killer. It wouldn't surprise me since I only have 512MB of RAM.

**Reply by 38github** [reply]
I am going to compile with less aggressive flags and see if NAM lv2 will use less RAM while not hurting the CPU efficiency.

**Reply by 38github** [reply]
For the moment I do not get any reconnects after compiling with the less aggressive flags/parameters and the CPU seem to stay the same. All I compile it with now is `-Ofast -flto -march=native`. Before that I used ridiculously many optimizations that I didn't really think through especially when considering this little SBC only has the mentioned 512MB of RAM.

**Reply by 38github** [reply]
I still get this message though:
![pipedal_nam_message](https://user-images.githubusercontent.com/18294149/235688586-7e247a69-e8de-464e-bf39-6699a25732bf.png)

**Reply by rerdavies** [reply]
Check that you have the most recent sources for NAM.

If you have the most recent sources, lines 142 to 146 of nam_plugin.cpp should read:

```
        void Plugin::process(uint32_t n_samples) noexcept
	{

		lv2_atom_forge_set_buffer(&atom_forge,(uint8_t*)ports.notify,ports.notify->atom.size);
		lv2_atom_forge_sequence_head(&atom_forge,&sequence_frame,uris.units_frame);
```

Original sources did fairly gruesome things with significant performance impacts.

**Reply by 38github** [reply]
I did a git pull but I am not on my laptop so I will look at those lines tomorrow.

**Reply by 38github** [reply]
I do not see those lines in nam_plugin.cpp. When was that added/changed?

**Reply by 38github** [reply]
Wait... I now noticed that I pulled NAM from your repo! I will immediatelly pull it from mikeoliphant and compile.

**Reply by 38github** [reply]
Thank you so much! The message is not appearing anymore!

**Reply by 38github** [reply]
I managed to half the RAM usage by changing the amount of http sockets in config.json from 10 to 2. Now the system use ~200 MiB of RAM instead of 380MiB. It has (please let it stay true) ended my issue with OOM killer killing Pipedal.

**Comment by rerdavies** [comment]
Well done! I guess I can imagine how that works. :-) I'll file that one away for future reference.

**Comment by hmollercl** [comment]
At least in my setups (ubuntu 24.04) mike oliphant NAM 0.1.9v3 (with x64 optimizations ) has a huge improvement in resource usage over ToobNAM. Maybe it happened after 0.1.7 release https://github.com/mikeoliphant/neural-amp-modeler-lv2/releases/tag/v0.1.7 not sure how it behaves in raspberry pi.
